jlm/llvm/ir/operators/MemCpy.hpp

```diff
@@ -0,0 +1,188 @@
+/*
+ * Copyright 2024 Nico Reißmann <nico.reissmann@gmail.com>
+ * See COPYING for terms of redistribution.
+ */
+
+#ifndef JLM_LLVM_IR_OPERATORS_MEMCPY_HPP
+#define JLM_LLVM_IR_OPERATORS_MEMCPY_HPP
+
+#include <jlm/llvm/ir/tac.hpp>
+#include <jlm/llvm/ir/types.hpp>
+#include <jlm/rvsdg/bitstring.hpp>
+#include <jlm/rvsdg/simple-node.hpp>
+
+namespace jlm::llvm
+{
+
+/**
+ * Represents an LLVM memcpy intrinsic.
+ *
+ * @see MemCpyVolatileOperation
+ */
+class MemCpyOperation final : public rvsdg::simple_op
+{
+public:
+  ~MemCpyOperation() override;
+
+  MemCpyOperation(
+      const std::vector<rvsdg::port> & operandPorts,
+      const std::vector<rvsdg::port> & resultPorts)
+      : simple_op(operandPorts, resultPorts)
+  {}
+
+  bool
+  operator==(const operation & other) const noexcept override;
+
+  [[nodiscard]] std::string
+  debug_string() const override;
+
+  [[nodiscard]] std::unique_ptr<rvsdg::operation>
+  copy() const override;
+
+  static std::unique_ptr<llvm::tac>
+  create(
+      const variable * destination,
+      const variable * source,
+      const variable * length,
+      const std::vector<const variable *> & memoryStates)
+  {
+    auto operandPorts = CheckAndCreateOperandPorts(length->type(), memoryStates.size());
+    auto resultPorts = CreateResultPorts(memoryStates.size());
+
+    std::vector<const variable *> operands = { destination, source, length };
+    operands.insert(operands.end(), memoryStates.begin(), memoryStates.end());
+
+    MemCpyOperation op(operandPorts, resultPorts);
+    return tac::create(op, operands);
```

Comment:
Would we save a copy by constructing the `MemCpyOperation` directly into the`tac::create()`? I never quite know when the C++ compiler elides calling copy constructors.

Reply:
Yes, we could save a copy here. We create the operation and then it is copied in the tac constructor:

```
tac::tac(const jlm::rvsdg::simple_op & operation, const std::vector<const variable *> & operands)
    : operands_(operands),
      operation_(operation.copy())
{
  check_operands(operation, operands);

  auto names = create_names(operation.nresults());
  create_results(operation, names);
}
```

However, I do not consider this change as part of this PR.